src/Bus/MongoBatchRepository.php

```diff
@@ -0,0 +1,276 @@
+        $batchId = new ObjectId($batchId);
+        $this->collection->updateOne(
+            ['_id' => $batchId],
+            [
+                '$inc' => [
+                    'total_jobs' => $amount,
+                    'pending_jobs' => $amount,
+                ],
+                '$set' => [
+                    'finished_at' => null,
+                ],
+            ],
+        );
+    }
+
+    #[Override]
+    public function decrementPendingJobs(string $batchId, string $jobId): UpdatedBatchJobCounts
+    {
+        $batchId = new ObjectId($batchId);
+        $values = $this->collection->findOneAndUpdate(
+            ['_id' => $batchId],
+            [
+                '$inc' => ['pending_jobs' => -1],
+                '$pull' => ['failed_job_ids' => $jobId],
+            ],
+            [
+                'projection' => ['pending_jobs' => 1, 'failed_jobs' => 1],
+                'returnDocument' => FindOneAndUpdate::RETURN_DOCUMENT_AFTER,
+            ],
+        );
+
+        return new UpdatedBatchJobCounts(
+            $values['pending_jobs'],
+            $values['failed_jobs'],
+        );
+    }
+
+    #[Override]
+    public function incrementFailedJobs(string $batchId, string $jobId): UpdatedBatchJobCounts
+    {
+        $batchId = new ObjectId($batchId);
+        $values = $this->collection->findOneAndUpdate(
+            ['_id' => $batchId],
+            [
+                '$inc' => ['failed_jobs' => 1],
+                '$push' => ['failed_job_ids' => $jobId],
+            ],
+            [
+                'projection' => ['pending_jobs' => 1, 'failed_jobs' => 1],
+                'returnDocument' => FindOneAndUpdate::RETURN_DOCUMENT_AFTER,
+            ],
+        );
+
+        return new UpdatedBatchJobCounts(
+            $values['pending_jobs'],
+            $values['failed_jobs'],
+        );
+    }
+
+    #[Override]
+    public function markAsFinished(string $batchId): void
+    {
+        $batchId = new ObjectId($batchId);
+        $this->collection->updateOne(
+            ['_id' => $batchId],
+            ['$set' => ['finished_at' => $this->getUTCDateTime()]],
+        );
+    }
+
+    #[Override]
+    public function cancel(string $batchId): void
+    {
+        $batchId = new ObjectId($batchId);
+        $this->collection->updateOne(
+            ['_id' => $batchId],
+            [
+                '$set' => [
+                    'cancelled_at' => $this->getUTCDateTime(),
+                    'finished_at' => $this->getUTCDateTime(),
+                ],
+            ],
+        );
+    }
+
+    #[Override]
+    public function delete(string $batchId): void
+    {
+        $batchId = new ObjectId($batchId);
+        $this->collection->deleteOne(['_id' => $batchId]);
+    }
+
+    /** Execute the given Closure within a storage specific transaction. */
+    #[Override]
+    public function transaction(Closure $callback): mixed
+    {
+        return $this->connection->transaction(fn () => $callback());
```

Comment:
Is there any significance to wrapping `$callback` in an additional closure?

Reply:
Removed. Also in Laravel from where I copied the code: https://github.com/laravel/framework/pull/51513